Build Fails with HIGHFIVE_PARALLEL_HDF5

I turned on HIGHFIVE_PARALLEL_HDF5 and now I'm getting build failures. My cmake output was 

> -- The C compiler identification is GNU 4.8.5
-- The CXX compiler identification is GNU 4.8.5
-- Check for working C compiler: /usr/tcs/gcc/4.8.5/bin/gcc
-- Check for working C compiler: /usr/tcs/gcc/4.8.5/bin/gcc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/tcs/gcc/4.8.5/bin/g++
-- Check for working CXX compiler: /usr/tcs/gcc/4.8.5/bin/g++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Setting build type to 'RelWithDebInfo' as none was specified.
-- Performing Test COMPILER_SUPPORTS_CXX11
-- Performing Test COMPILER_SUPPORTS_CXX11 - Success
-- Performing Test COMPILER_SUPPORTS_CXX14
-- Performing Test COMPILER_SUPPORTS_CXX14 - Success
-- Performing Test COMPILER_SUPPORTS_CXX17
-- Performing Test COMPILER_SUPPORTS_CXX17 - Success
-- HDF5: Using hdf5 compiler wrapper to determine C configuration
-- Found HDF5: /home/tee31163/anaconda3/lib/libhdf5.so;/usr/lib64/librt.so;/usr/lib64/libpthread.so;/home/tee31163/anaconda3/lib/libz.so;/usr/lib64/libdl.so;/usr/lib64/libm.so (found version "1.10.4")
-- Boost version: 1.53.0
-- Found the following Boost libraries:
--   system
--   serialization
--   unit_test_framework
-- Found MPI_C: /usr/lib64/openmpi/lib/libmpi.so (found version "3.0")
-- Found MPI_CXX: /usr/lib64/openmpi/lib/libmpi_cxx.so (found version "3.0")
-- Found MPI: TRUE (found version "3.0")
--  - AUTO_TEST_WITH_MPIEXEC cmd /usr/lib64/openmpi/bin/mpiexec
--   -- set MPI test execution prefix to /usr/lib64/openmpi/bin/mpiexec
-- Found Doxygen: /usr/bin/doxygen (found version "1.8.5") found components:  doxygen dot
-- Configuring done
-- Generating done
-- Build files have been written to: /home/tee31163/HighFive/build

The verbose make output is:

> /home/tee31163/anaconda3/bin/cmake -S/home/tee31163/HighFive -B/home/tee31163/HighFive/build --check-build-system CMakeFiles/Makefile.cmake 0
/home/tee31163/anaconda3/bin/cmake -E cmake_progress_start /home/tee31163/HighFive/build/CMakeFiles /home/tee31163/HighFive/build/CMakeFiles/progress.marks
make -f CMakeFiles/Makefile2 all
make[1]: Entering directory `/home/tee31163/HighFive/build'
make -f src/examples/CMakeFiles/select_partial_dataset_cpp11_bin.dir/build.make src/examples/CMakeFiles/select_partial_dataset_cpp11_bin.dir/depend
make[2]: Entering directory `/home/tee31163/HighFive/build'
cd /home/tee31163/HighFive/build && /home/tee31163/anaconda3/bin/cmake -E cmake_depends "Unix Makefiles" /home/tee31163/HighFive /home/tee31163/HighFive/src/examples /home/tee31163/HighFive/build /home/tee31163/HighFive/build/src/examples /home/tee31163/HighFive/build/src/examples/CMakeFiles/select_partial_dataset_cpp11_bin.dir/DependInfo.cmake --color=
Dependee "/home/tee31163/HighFive/build/src/examples/CMakeFiles/select_partial_dataset_cpp11_bin.dir/DependInfo.cmake" is newer than depender "/home/tee31163/HighFive/build/src/examples/CMakeFiles/select_partial_dataset_cpp11_bin.dir/depend.internal".
Dependee "/home/tee31163/HighFive/build/src/examples/CMakeFiles/CMakeDirectoryInformation.cmake" is newer than depender "/home/tee31163/HighFive/build/src/examples/CMakeFiles/select_partial_dataset_cpp11_bin.dir/depend.internal".
Scanning dependencies of target select_partial_dataset_cpp11_bin
make[2]: Leaving directory `/home/tee31163/HighFive/build'
make -f src/examples/CMakeFiles/select_partial_dataset_cpp11_bin.dir/build.make src/examples/CMakeFiles/select_partial_dataset_cpp11_bin.dir/build
make[2]: Entering directory `/home/tee31163/HighFive/build'
[  3%] Building CXX object src/examples/CMakeFiles/select_partial_dataset_cpp11_bin.dir/select_partial_dataset_cpp11.cpp.o
cd /home/tee31163/HighFive/build/src/examples && /usr/tcs/gcc/4.8.5/bin/g++  -DH5_USE_BOOST -DMPI_NO_CPPBIND -DNDEBUG -D_FORTIFY_SOURCE=2 -I/home/tee31163/HighFive/include -isystem /home/tee31163/anaconda3/include -isystem /usr/include/openmpi-x86_64  -g -Wall -Wextra -Werror -Wshadow -Wnon-virtual-dtor -Wcast-align -Wunused -Woverloaded-virtual -Wpedantic -Wdouble-promotion -Wformat=2 -Wconversion -Wsign-conversion -Wno-unused-parameter -O2   -std=gnu++1y -o CMakeFiles/select_partial_dataset_cpp11_bin.dir/select_partial_dataset_cpp11.cpp.o -c /home/tee31163/HighFive/src/examples/select_partial_dataset_cpp11.cpp
[  6%] Linking CXX executable select_partial_dataset_cpp11_bin
cd /home/tee31163/HighFive/build/src/examples && /home/tee31163/anaconda3/bin/cmake -E cmake_link_script CMakeFiles/select_partial_dataset_cpp11_bin.dir/link.txt --verbose=1
/usr/tcs/gcc/4.8.5/bin/g++  -g -Wall -Wextra -Werror -Wshadow -Wnon-virtual-dtor -Wcast-align -Wunused -Woverloaded-virtual -Wpedantic -Wdouble-promotion -Wformat=2 -Wconversion -Wsign-conversion -Wno-unused-parameter -O2  "-Wl,-rpath -Wl,/usr/lib64/openmpi/lib -Wl,--enable-new-dtags -pthread" CMakeFiles/select_partial_dataset_cpp11_bin.dir/select_partial_dataset_cpp11.cpp.o  -o select_partial_dataset_cpp11_bin -Wl,-rpath,/home/tee31163/anaconda3/lib:/usr/lib64/openmpi/lib /home/tee31163/anaconda3/lib/libhdf5.so /usr/lib64/librt.so -lpthread /home/tee31163/anaconda3/lib/libz.so /usr/lib64/libdl.so /usr/lib64/libm.so /usr/lib64/openmpi/lib/libmpi.so /usr/lib64/libboost_serialization-mt.so
/usr/bin/ld: unrecognised option: -rpath -Wl
collect2: error: ld returned 1 exit status
make[2]: *** [src/examples/select_partial_dataset_cpp11_bin] Error 1
make[2]: Leaving directory `/home/tee31163/HighFive/build'
make[1]: *** [src/examples/CMakeFiles/select_partial_dataset_cpp11_bin.dir/all] Error 2
make[1]: Leaving directory `/home/tee31163/HighFive/build'
make: *** [all] Error 2

This was with a recently cloned repo. The only modification I made was turning on parallel builds. It worked before making this change.

--- a/.travis.yml
+++ b/.travis.yml
@@ -23,7 +23,7 @@ addons: &clang6
     packages:
       - clang-6.0
       - libboost-all-dev
-      - libhdf5-mpich-dev
+      - libhdf5-openmpi-dev
       - libeigen3-dev
       - ninja-build
     sources:

--- a/.travis.yml
+++ b/.travis.yml
@@ -23,7 +23,7 @@ addons: &clang6
     packages:
       - clang-6.0
       - libboost-all-dev
-      - libhdf5-openmpi-dev
+      - libhdf5-mpich-dev
       - libeigen3-dev
       - ninja-build
     sources:

--- a/.travis.yml
+++ b/.travis.yml
@@ -86,7 +86,7 @@ matrix:
     # osx
     - env:
       - USE_XTENSOR=True
-      - HIGHFIVE_PARALLEL_HDF5=False
+      - HIGHFIVE_PARALLEL_HDF5=True
       os: osx
       osx_image: xcode10.3
 
@@ -105,7 +105,7 @@ install:
   - if [[ "$TRAVIS_OS_NAME" == "linux" ]]; then
       MINCONDA_OS=$MINCONDA_LINUX;
     elif [[ "$TRAVIS_OS_NAME" == "osx" ]]; then
-      HOMEBREW_NO_AUTO_UPDATE=1 brew install boost hdf5 eigen ninja;
+      HOMEBREW_NO_AUTO_UPDATE=1 brew install boost hdf5 --with-mpi eigen ninja;
       MINCONDA_OS=$MINCONDA_OSX;
     fi
 

--- a/.travis.yml
+++ b/.travis.yml
@@ -44,56 +44,62 @@ addons: &clang7
 
 matrix:
   include:
-    # Test default gcc on Ubuntu trusty
-    - env:
-      - USE_XTENSOR=False
-      - HIGHFIVE_PARALLEL_HDF5=True
-      os: linux
+    # Older linux (trusty) with default gcc
+    # Install serial hdf5 + build serial
+    - os: linux
       dist: trusty
+      env:
+        - USE_XTENSOR=False
+        - HIGHFIVE_PARALLEL_HDF5=False
       addons:
         apt:
           packages:
-          - libboost-all-dev
-          - libeigen3-dev
-          - libhdf5-openmpi-dev
-          - ninja-build
-    # Test gcc-7
-    - env:
-      - GCC_VERSION=7
-      - USE_XTENSOR=True
-      - HIGHFIVE_PARALLEL_HDF5=True
-      os: linux
+            - libboost-all-dev
+            - libeigen3-dev
+            - libhdf5-serial-dev
+            - ninja-build
+
+    # Linux gcc-7
+    # Install parallel hdf5 + build parallel
+    - os: linux
       dist: xenial
+      env:
+        - GCC_VERSION=7
+        - USE_XTENSOR=True
+        - HIGHFIVE_PARALLEL_HDF5=True
       addons: *gcc7
 
-    # Test clang-6.0
-    - env:
-      - CLANG_VERSION=6.0
-      - HIGHFIVE_PARALLEL_HDF5=True
-      os: linux
+    # Linux clang-6.0
+    # Install parallel hdf5 + build serial
+    - os: linux
       dist: xenial
+      env:
+        - CLANG_VERSION=6.0
+        - HIGHFIVE_PARALLEL_HDF5=False
       addons: *clang6
 
-    # Test clang-7
-    - env:
-      - CLANG_VERSION=7
-      - USE_XTENSOR=True
-      - HIGHFIVE_PARALLEL_HDF5=True
-      os: linux
+    # Linux clang-7
+    # Install parallel hdf5, build parallel
+    - os: linux
       dist: xenial
+      env:
+        - CLANG_VERSION=7
+        - USE_XTENSOR=True
+        - HIGHFIVE_PARALLEL_HDF5=True
       addons: *clang7
 
-    # osx
-    - env:
-      - USE_XTENSOR=True
-      - HIGHFIVE_PARALLEL_HDF5=False
-      os: osx
+    # Mac OSX
+    - os: osx
       osx_image: xcode10.3
+      env:
+        - USE_XTENSOR=True
+        - HIGHFIVE_PARALLEL_HDF5=False
 
-    - env:
-      - USE_XTENSOR=True
-      - HIGHFIVE_PARALLEL_HDF5=False
-      os: windows
+    # Windows
+    - os: windows
+      env:
+        - USE_XTENSOR=True
+        - HIGHFIVE_PARALLEL_HDF5=False
 
 env:
   global:

--- a/.travis.yml
+++ b/.travis.yml
@@ -12,7 +12,7 @@ addons: &gcc7
     packages:
       - g++-7
       - libboost-all-dev
-      - libhdf5-serial-dev
+      - libhdf5-openmpi-dev
       - libeigen3-dev
       - ninja-build
     sources:
@@ -23,7 +23,7 @@ addons: &clang6
     packages:
       - clang-6.0
       - libboost-all-dev
-      - libhdf5-serial-dev
+      - libhdf5-openmpi-dev
       - libeigen3-dev
       - ninja-build
     sources:
@@ -35,7 +35,7 @@ addons: &clang7
     packages:
       - clang-7
       - libboost-all-dev
-      - libhdf5-serial-dev
+      - libhdf5-openmpi-dev
       - libeigen3-dev
       - ninja-build
     sources:
@@ -47,6 +47,7 @@ matrix:
     # Test default gcc on Ubuntu trusty
     - env:
       - USE_XTENSOR=False
+      - HIGHFIVE_PARALLEL_HDF5=True
       os: linux
       dist: trusty
       addons:
@@ -54,12 +55,13 @@ matrix:
           packages:
           - libboost-all-dev
           - libeigen3-dev
-          - libhdf5-serial-dev
+          - libhdf5-openmpi-dev
           - ninja-build
     # Test gcc-7
     - env:
       - GCC_VERSION=7
       - USE_XTENSOR=True
+      - HIGHFIVE_PARALLEL_HDF5=True
       os: linux
       dist: xenial
       addons: *gcc7
@@ -67,6 +69,7 @@ matrix:
     # Test clang-6.0
     - env:
       - CLANG_VERSION=6.0
+      - HIGHFIVE_PARALLEL_HDF5=True
       os: linux
       dist: xenial
       addons: *clang6
@@ -75,6 +78,7 @@ matrix:
     - env:
       - CLANG_VERSION=7
       - USE_XTENSOR=True
+      - HIGHFIVE_PARALLEL_HDF5=True
       os: linux
       dist: xenial
       addons: *clang7
@@ -82,10 +86,14 @@ matrix:
     # osx
     - env:
       - USE_XTENSOR=True
+      - HIGHFIVE_PARALLEL_HDF5=False
       os: osx
       osx_image: xcode10.3
 
-    - os: windows
+    - env:
+      - USE_XTENSOR=True
+      - HIGHFIVE_PARALLEL_HDF5=False
+      os: windows
 
 env:
   global:
@@ -137,6 +145,7 @@ script:
     cmake --warn-uninitialized --debug-output
     -DCMAKE_VERBOSE_MAKEFILE:BOOL=ON
     -DHIGHFIVE_TEST_SINGLE_INCLUDES:BOOL=ON
+    -DHIGHFIVE_PARALLEL_HDF5:BOOL=${HIGHFIVE_PARALLEL_HDF5}
     -DUSE_EIGEN:BOOL=ON
     -DUSE_XTENSOR:BOOL=${USE_XTENSOR}
     -G "${CMAKE_GENERATOR}" ../

--- a/.travis.yml
+++ b/.travis.yml
@@ -86,7 +86,7 @@ matrix:
     # osx
     - env:
       - USE_XTENSOR=True
-      - HIGHFIVE_PARALLEL_HDF5=True
+      - HIGHFIVE_PARALLEL_HDF5=False
       os: osx
       osx_image: xcode10.3
 
@@ -105,7 +105,7 @@ install:
   - if [[ "$TRAVIS_OS_NAME" == "linux" ]]; then
       MINCONDA_OS=$MINCONDA_LINUX;
     elif [[ "$TRAVIS_OS_NAME" == "osx" ]]; then
-      HOMEBREW_NO_AUTO_UPDATE=1 brew install boost hdf5 --with-mpi eigen ninja;
+      HOMEBREW_NO_AUTO_UPDATE=1 brew install boost hdf5 eigen ninja;
       MINCONDA_OS=$MINCONDA_OSX;
     fi
 

--- a/CMake/PackageConfig.cmake
+++ b/CMake/PackageConfig.cmake
@@ -10,12 +10,12 @@ target_compile_definitions(HighFive INTERFACE ${HDF5_DEFINITIONS})
 
 # MPI
 if(HIGHFIVE_PARALLEL_HDF5)
-  target_include_directories(HighFive SYSTEM INTERFACE ${MPI_C_INCLUDE_PATH})
-  target_link_libraries(HighFive INTERFACE ${MPI_C_LIBRARIES})
+  target_include_directories(HighFive SYSTEM INTERFACE ${MPI_CXX_INCLUDE_PATH})
+  target_link_libraries(HighFive INTERFACE ${MPI_CXX_LIBRARIES})
   if(CMAKE_VERSION VERSION_LESS 3.13)
-    target_link_libraries(HighFive INTERFACE ${MPI_C_LINK_FLAGS})
+    target_link_libraries(HighFive INTERFACE ${MPI_CXX_LINK_FLAGS})
   else()
-    target_link_options(HighFive INTERFACE ${MPI_C_LINK_FLAGS})
+    target_link_options(HighFive INTERFACE "SHELL:${MPI_CXX_LINK_FLAGS}")
   endif()
 endif()
 

--- a/CMake/PackageConfig.cmake
+++ b/CMake/PackageConfig.cmake
@@ -9,7 +9,7 @@ target_link_libraries(HighFive INTERFACE ${HDF5_LIBRARIES})
 target_compile_definitions(HighFive INTERFACE ${HDF5_DEFINITIONS})
 
 # MPI
-if(HIGHFIVE_PARALLEL_HDF5)
+if(HIGHFIVE_PARALLEL_HDF5 OR HDF5_IS_PARALLEL)
   target_include_directories(HighFive SYSTEM INTERFACE ${MPI_CXX_INCLUDE_PATH})
   target_link_libraries(HighFive INTERFACE ${MPI_CXX_LIBRARIES})
   if(CMAKE_VERSION VERSION_LESS 3.13)

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -78,7 +78,7 @@ if(USE_XTENSOR)
   endif()
 endif()
 
-if(HIGHFIVE_PARALLEL_HDF5)
+if(HIGHFIVE_PARALLEL_HDF5 OR HDF5_IS_PARALLEL)
   find_package(MPI REQUIRED)
   add_definitions("-DMPI_NO_CPPBIND")
 endif()

